In [23]:
import time

available_coins = [50, 25, 10, 5, 2, 1]

def find_coins_greedy(amount):
    coins = sorted(available_coins, reverse=True)
    result = {}
    iterations = 0
    start_time = time.time()
    amount_left = amount
    while amount_left > 0:
        for coin in coins:
            if amount_left - coin >= 0:
                if coin in result:
                    result[coin] += 1
                else:
                    result[coin] = 1
                amount_left -= coin
                iterations += 1
                break
    end_time = time.time()
    execution_time = end_time - start_time
    return result, iterations, execution_time

def find_coins_greedy_optimized(amount):
    coins = sorted(available_coins, reverse=True)
    result = {}
    iterations = 0
    start_time = time.time()
    for coin in coins:
        count = amount // coin
        iterations += 1
        if count > 0:
            result[coin] = count
            amount -= count * coin
    end_time = time.time()
    execution_time = end_time - start_time
    return result, iterations, execution_time

def find_coins_dp(amount):
    start_time = time.time()
    coins = sorted(available_coins, reverse=True)
    dp = [float('inf')] * (amount + 1)
    dp[0] = 0
    iterations = 0
    coin_used = [0] * (amount + 1)
    for i in range(1, amount + 1):
        for coin in coins:
            if i >= coin and dp[i - coin] + 1 < dp[i]:
                dp[i] = dp[i - coin] + 1
                coin_used[i] = coin
                iterations += 1
    
    result = {}
    while amount > 0:
        count = amount // coin_used[amount]
        if count > 0:
            result[coin_used[amount]] = count
            amount -= count * coin_used[amount]
    
    end_time = time.time()
    execution_time = end_time - start_time
    return result, iterations, execution_time


# Приклад використання
amount = 654654
greedy_result, greedy_iterations, greedy_time = find_coins_greedy(amount)
min_coins_result, min_coins_iterations, min_coins_time = find_coins_dp(amount)
opt_greedy_result, opt_greedy_iterations, opt_greedy_time = find_coins_greedy_optimized(amount)

print("Жадібний алгоритм:")
print("Результат:", greedy_result)
print("Кількість ітерацій:", greedy_iterations)
print("Час виконання:", greedy_time)

print("\nДинамічне програмування:")
print("Результат:", min_coins_result)
print("Кількість ітерацій:", min_coins_iterations)
print("Час виконання:", min_coins_time)


print("\nОптимізований Жадібний алгоритм:")
print("Результат:", opt_greedy_result)
print("Кількість ітерацій:", opt_greedy_iterations)
print("Час виконання:", opt_greedy_time)


Жадібний алгоритм:
Результат: {50: 13093, 2: 2}
Кількість ітерацій: 13095
Час виконання: 0.004462242126464844

Динамічне програмування:
Результат: {50: 13093, 2: 2}
Кількість ітерацій: 654654
Час виконання: 0.5262265205383301

Оптимізований Жадібний алгоритм:
Результат: {50: 13093, 2: 2}
Кількість ітерацій: 6
Час виконання: 1.9073486328125e-06


# Опис

Було порівняно Жадібний алгоритм та Алгоритм динамічного програмування

# Результат

| Алгоритм | Результат                | Кількість ітерацій | Час виконання     |
|----------|--------------------------|--------------------|-------------------|
| Жадібний | {50: 13093, 2: 2}       | 13095              | 0.004462242126464844 |
| Динамічне програмування | {50: 13093, 2: 2} | 654654             | 0.5262265205383301  |
| Оптимізований Жадібний | {50: 13093, 2: 2} | 6                  | 1.9073486328125e-06 |


# Висновок

Як видно з результатів з такою задачею краще працює жадібний алгоритм. Також якщо оптимізувати такий алгортим, можна отримати приріст в часі виконання з 4.4 мілісекунд до 1.9 мікросекунд. Динамічні алгоритми з цією задачею справляються значно повільніше 